python/revng/ptml/actions.py

```diff
@@ -0,0 +1,97 @@
+
+    @staticmethod
+    def parse(string: str) -> Optional["Location"]:
+        if not string.startswith("/"):
+            return None
+
+        parts = string.split("/")[1:]
+        return Location(parts[0], parts[1:])
+
+    def to_string(self) -> str:
+        return f"/{self.type}/{'/'.join(self.path_components)}"
+
+
+@dataclass(frozen=True)
+class ActionMapping:
+    # Given a PipelineLocation, return the TupleTreePath for the model
+    rename: Optional[Callable[[Location], str]]
+    # Given a PipelineLocation, return the TupleTreePath for the model
+    comment: Optional[Callable[[Location], str]]
+
+
+def _model_path_generator(*args: Union[str, Tuple[int, bool]]) -> Callable[[Location], str]:
+    def generator(location: Location):
+        if len(args) == 0:
+            return ""
+
+        result = ""
+        for element in args:
+            if not result.endswith("::"):
+                result += "/"
+
+            if isinstance(element, str):
+                result += element
+            else:
+                if element[1]:
+                    # Path needs to be upcasted
+                    part = location.path_components[element[0]]
+                    kind = part.split("-")[-1]
+                    result += f"{part}/{kind}::"
+                else:
+                    result += location.path_components[element[0]]
+        return result
+
+    return generator
+
+
+def _common(*args: Union[str, Tuple[int, bool]]) -> ActionMapping:
+    return ActionMapping(
+        _model_path_generator(*args, "CustomName"),
+        _model_path_generator(*args, "Comment"),
+    )
+
+
+ACTION_MAPPING: Dict[str, ActionMapping] = {
+    "function": _common("Functions", (0, False)),
+    "type-definition": _common("TypeDefinitions", (0, True)),
+    "struct-field": _common("TypeDefinitions", (0, True), "Fields", (1, False)),
+    "union-field": _common("TypeDefinitions", (0, True), "Fields", (1, False)),
+    "enum-entry": _common("TypeDefinitions", (0, True), "Entries", (1, False)),
+    "cabi-argument": _common("TypeDefinitions", (0, True), "Arguments", (1, False)),
+    "raw-argument": _common("TypeDefinitions", (0, True), "Arguments", (1, False)),
+    "return-value": ActionMapping(
+        None, _model_path_generator("TypeDefinitions", (0, True), "ReturnValueComment")
+    ),
+    "return-register": _common("TypeDefinitions", (0, True), "ReturnValues", (1, False)),
+    "segment": _common("Segments", (0, False)),
+    "dynamic-function": _common("ImportedDynamicFunctions", (0, False)),
+}
+
+
+def rename(location: Location) -> Optional[str]:
+    mapping = ACTION_MAPPING.get(location.type)
+    if mapping is None or mapping.rename is None:
+        return None
+    return mapping.rename(location)
+
+
+def comment(location: Location) -> Optional[str]:
+    mapping = ACTION_MAPPING.get(location.type)
+    if mapping is None or mapping.comment is None:
+        return None
+    return mapping.comment(location)
```

Comment:
It is unclear what these functions do. They do not actually *rename* things. They produce a model path to change, AFAIU.

Maybe `model_path_for_{rename,comment}`?